common/lockfile/src/lib.rs

```diff
@@ -0,0 +1,38 @@
+use fs2::FileExt;
+use std::fs::{self, File};
+use std::io;
+use std::path::PathBuf;
+
+/// Cross-platform file lock that auto-deletes on drop.
+///
+/// This lockfile uses OS locking primitives (`flock` on Unix, `LockFile` on Windows), and will
+/// only fail if locked by another process. I.e. if the file being locked already exists but isn't
+/// locked, then it can still be locked. This is relevant if an ungraceful shutdown (SIGKILL, power
+/// outage) caused the lockfile not to be deleted.
+#[derive(Debug)]
+pub struct Lockfile {
+    file: File,
+    path: PathBuf,
+}
+
+#[derive(Debug)]
+pub enum LockfileError {
+    FileLocked(PathBuf, io::Error),
+    UnableToCreateFile(io::Error),
+}
+
+impl Lockfile {
+    /// Obtain an exclusive lock on the file at `path`, creating it if it doesn't exist.
+    pub fn new(path: PathBuf) -> Result<Self, LockfileError> {
+        let file = File::create(&path).map_err(LockfileError::UnableToCreateFile)?;
+        file.try_lock_exclusive()
+            .map_err(|e| LockfileError::FileLocked(path.clone(), e))?;
+        Ok(Self { file, path })
+    }
+}
+
+impl Drop for Lockfile {
+    fn drop(&mut self) {
+        let _ = fs::remove_file(&self.path);
```

Comment:
Ah that's a good point. I know closing the file descriptor does, which happens shortly after this call anyway, but it might be prudent to call `unlock` as well. I'll investigate.

Reply:
I've decided not to change anything here for now, in the interests of keeping things simple and not introducing more potential race conditions. I realised if we added an `unlock` before the delete, then the following sequence would be possible:

```
1. Process A unlocks file
2. Process B locks file
3. Process A deletes file
4. ???
```